cedar-policy-core/src/entities/conformance.rs

```diff
@@ -172,109 +171,158 @@ pub fn typecheck_value_against_schematype(
+                        if !v.required {
+                            Ok(())
+                        } else {
+                            match pairs_map.get(k) {
+                                Some(inner_e) => typecheck_restricted_expr_against_schematype(
+                                    *inner_e,
+                                    &v.attr_type,
+                                    extensions,
+                                ),
+                                None => Err(TypeMismatchError::missing_required_attr(
+                                    expected_ty.clone(),
+                                    k.clone(),
+                                    expr.to_owned(),
+                                )
+                                .into()),
+                            }
                         }
-                    }
-                }))?;
-                let all_rec_attrs_match_schema =
-                    try_all(pairs_map.iter().map(|(k, inner_e)| match attrs.get(*k) {
-                        Some(sch_ty) => does_restricted_expr_implement_schematype(
+                    })
+                    .collect::<Result<(), _>>()?;
+                // Check that all attributes in the record are present (as
+                // required or optional) in the schema.
+                pairs_map
+                    .iter()
+                    .map(|(k, inner_e)| match attrs.get(*k) {
+                        Some(sch_ty) => typecheck_restricted_expr_against_schematype(
                             *inner_e,
                             &sch_ty.attr_type,
                             extensions,
                         ),
-                        None => Ok(*open_attrs),
-                    }))?;
-                all_rec_attrs_match_schema && all_req_schema_attrs_in_record
+                        None => {
+                            if *open_attrs {
+                                Ok(())
+                            } else {
+                                Err(TypeMismatchError::unexpected_attr(
+                                    expected_ty.clone(),
+                                    (*k).clone(),
+                                    expr.to_owned(),
+                                )
+                                .into())
+                            }
+                        }
+                    })
+                    .collect::<Result<(), _>>()?;
+                Ok(())
             }
-            None => false,
-        },
-        Extension { name } => match expr.as_extn_fn_call() {
-            Some((actual_name, _)) => match name.0.id.as_ref() {
-                "ipaddr" => actual_name.0.id.as_ref() == "ip",
-                _ => name == actual_name,
-            },
-            None => false,
+            None => type_mismatch_err(),
         },
+        // Extension functions are handled by the first `match` in this function.
+        Extension { .. } => type_mismatch_err(),
```

Comment:
nit: in the first `match`, I think we only handled the case where we expected an extension type.  What if we are expecting a builtin type, and the restrictedexpr is an extension function that returns that type?  (Pretty sure this case is not reachable from public API surface so it kinda doesn't matter, but)

Reply:
I think I actually fixed the behavior on extension functions with the changes in this PR. The first `match` compares the return type to the expected type instead of looking at the function name.

Reply:
also added some test cases hitting this